### ISSUE TITLE
Annotation processor infinite recursion on self-referencing generics

### ISSUE THREAD

**Issue by Pyrofab** [SpongePowered/Mixin#384]
Using yarn mappings:

```java
@Mixin(GameRules.class)
public interface GameRulesAccessor {
    @Invoker
    @SuppressWarnings("PublicStaticMixinMember")
    static <T extends GameRules.Rule<T>> GameRules.RuleKey<T> invokeRegister(String name, GameRules.RuleType<T> type) {
        throw new AssertionError();
    }
}
```

Throws a StackOverflowError with the following stacktrace : https://gist.github.com/Pyrofab/5b866026885232d0e7952f144cb731ff

**Comment by Mumfrey** [comment]
Hmmm ok, I'll look into that. I'm not sure how to calculate the upper bound there so it might just have to be caught as a corner-case.

**Comment by Mumfrey** [comment]
Can no longer replicate this with the example provided so I assume it was fixed in 0.8 somewhere along the line.
